File: VethelpAdmin/invoice.cs
using System;
using System.Collections.Generic;
using System.IO;
using Newtonsoft.Json;
using Newtonsoft.Json.Serialization;



//igualar os campos dos atributos da classe com os nomes dados no codigo python
// Id = id
public class Invoice
{
    public int Id { get; set; }
    public int Recepcionist_id { get; set; }
    public int Vet_id { get; set; }
    public int Animal_id { get; set; }
    public int Procedure_id { get; set; }
    public int Product_id { get; set; }
    public string Description { get; set; }
    public string Date { get; set; }
    public int Status { get; set; }

    // Construtor para inicializar os atributos
    public Invoice(int id, int recepcionist_id, int vet_id, int animal_id, int procedure_id, int product_id, string description, string date, int status)
    {
        Id = id;
        Recepcionist_id = recepcionist_id;
        Vet_id = vet_id;
        Animal_id = animal_id;
        Procedure_id = procedure_id;
        Product_id = product_id;
        Description = description;
        Date = date;
        Status = status;
    }

    // Método que vai retornar uma string representando o objeto
    public override string ToString()
    {
        return $"{Id} - {Recepcionist_id} - {Vet_id} - {Animal_id} - {Procedure_id} - {Product_id} - {Description} - {Date} - {Status}";
    }
}

public class Invoices
{
    // Lista de objetos do tipo Invoice
    private static List<Invoice> objetos = new List<Invoice>();
    private static string _filePath = "invoices.json"; // Nome do arquivo
    private static JsonSerializerSettings _jsonSettings = new JsonSerializerSettings
    {
        ContractResolver = new CamelCasePropertyNamesContractResolver() // Converte para camelCase (minúsculo)
    };


    // Método para inserir um novo usuário
    public static void Inserir(Invoice obj)
    {
        Abrir(); // Abre o arquivo para carregar os dados
        // Calcula o ID do novo objeto, incrementando o ID máximo já existente
        int id = 0;
        foreach (var x in objetos)
        {
            if (x.Id > id) id = x.Id;
        }
        obj.Id = id + 1; // Atribui um novo ID ao objeto
        objetos.Add(obj); // Adiciona o novo objeto à lista
        Salvar(); // Salva a lista de usuários no arquivo JSON
    }

    // Método para listar todos os usuários
    public static List<Invoice> Listar()
    {
        Abrir(); // Carrega a lista de objetos
        return objetos; // Retorna a lista de usuários
    }

    // Método para listar um usuário pelo ID
    public static Invoice ListarId(int id)
    {
        Abrir(); // Carrega a lista de objetos
        return objetos.Find(x => x.Id == id); // Retorna o usuário com o ID correspondente
    }

    // Método para atualizar um usuário
    public static void Atualizar(Invoice obj)
    {
        var usuario = ListarId(obj.Id); // Busca o usuário pelo ID
        if (usuario != null)
        {
            objetos.Remove(usuario); // Remove o usuário antigo
            objetos.Add(obj); // Adiciona o usuário atualizado
            Salvar(); // Salva as alterações no arquivo JSON
        }
    }

    // Método para excluir um usuário
    public static void Excluir(Invoice obj)
    {
        var usuario = ListarId(obj.Id); // Busca o usuário pelo ID
        if (usuario != null)
        {
            objetos.Remove(usuario); // Remove o usuário
            Salvar(); // Salva as alterações no arquivo JSON
        }
    }

    // Método para salvar os dados no arquivo JSON
    private static void Salvar()
    {
        string json = JsonConvert.SerializeObject(objetos, Formatting.Indented, _jsonSettings); // Usa as configurações

        string basePath = Directory.GetCurrentDirectory(); // Pasta atual
        string vethelpPath = Path.Combine(basePath, "..", _filePath); // Volta um diretório

        File.WriteAllText(vethelpPath, json); // Escreve o JSON no arquivo no caminho correto
    
    }

    // Método para carregar os dados do arquivo JSON
    private static void Abrir()
    {
        objetos.Clear(); // Limpa a lista de objetos
        string basePath = Directory.GetCurrentDirectory();
        string vethelpPath = Path.Combine(basePath, "..", _filePath); // Volta um diretório

        if (File.Exists(vethelpPath)) // Verifica se o arquivo existe no caminho correto
        {
            string json = File.ReadAllText(vethelpPath); // Lê o arquivo JSON do caminho correto
            objetos = JsonConvert.DeserializeObject<List<Invoice>>(json, _jsonSettings); // Usa as configurações na desserialização

        }
    }
}

File: models/invoice.py
import json
import datetime

class Invoice:
    def __init__(self, id, data, id_service, status, total = 0.0):
        self.id = id # atributos de instância
        self.data = data
        self.id_service = id_service
        self.status = status


    def __str__(self):
        return f"Id: {self.id} - Data: {self.data} - Total: R${self.total:.2f}"

class Invoices:
    objetos = [] # atributo de classe
    @classmethod
    def criar(cls, obj):
        # abre a lista do arquivo
        cls.abrir()
        # calcula o id do objeto
        id = 0
        for x in cls.objetos:
            if x.id > id: id = x.id
        obj.id = id + 1    
        # insere o objeto na lista
        cls.objetos.append(obj)
        # salva a lista no arquivo
        cls.salvar()
    @classmethod
    def listar(cls):
        # abre a lista do arquivo
        cls.abrir()
        # retorna a lista para a UI
        return cls.objetos[:]
    @classmethod
    def listar_id(cls, id):
        cls.abrir()
        # percorre a lista procurando o id    
        for x in cls.objetos:
            if x.id == id: return x
        return None
    @classmethod
    def atualizar(cls, obj):
        x = cls.listar_id(obj.id)
        if x != None:
            cls.objetos.remove(x)
            cls.objetos.append(obj)
            cls.salvar()        
    @classmethod
    def excluir(cls, obj):
        x = cls.listar_id(obj.id)
        if x != None:
            cls.objetos.remove(x)
            cls.salvar()
    
    @classmethod
    def fechar(cls, obj):
        x = cls.listar_id(obj.id)
        if x != None:
            x.status = False
            cls.salvar()

    @classmethod
    def atualizar_total(cls, invoice, valor):
        x = cls.listar_id(invoice.id)
        if x != None:
            x.total = x.total + valor
            cls.salvar()
    
    @classmethod
    def mostrar_total(cls, id):
        x = cls.listar(id)
        if x != None:
            return x.total
        

    @classmethod
    def salvar(cls):
        # open - cria e abre o arquivo invoices.json
        # vars - converte um objeto em um dicionário
        # dump - pega a lista de objetos e salva no arquivo
        with open("invoices.json", mode="w") as arquivo:
            json.dump(cls.objetos, arquivo, default = vars)
    @classmethod
    def abrir(cls):
        # esvazia a lista de objetos
        cls.objetos = []
        try:
            with open("invoices.json", mode="r") as arquivo:
                # abre o arquivo com a lista de dicionários -> clientes_json
                objetos_json = json.load(arquivo)
                # percorre a lista de dicionários
                for obj in objetos_json:
                    # recupera cada dicionário e cria um objeto
                    c = Invoice(obj["id"], obj["data"], obj["carrinho"], obj["total"], obj["id_cliente"])
                    # insere o objeto na lista
                    cls.objetos.append(c)    
        except FileNotFoundError:
            pass
    

